File: src/ch01/meetup/go/higher_order.go
package main

import "fmt"

type Unary func(float64) float64

func main() {
	compose := func(f Unary, g Unary) Unary {
		return Unary(func(x float64) float64 {
			return f(g(x))
		})
	}

	make_adder := func(n float64) Unary {
		return Unary(func(x float64) float64 {
			return x + n
		})
	}

	make_multiplier := func(n float64) Unary {
		return Unary(func(x float64) float64 {
			return x * n
		})
	}

	fmt.Println(compose(make_multiplier(3), make_adder(1))(10))
}

File: src/ch01/meetup/scala/higher_order.scala
object Main extends App {
  def makeAdder(n: Int): Int => Int = (x: Int) => x + n
  def makeMultiplier(n: Int): Int => Int = (x: Int) => x * n
  def compose(f: Int => Int, g: Int => Int): Int => Int = (x: Int) => f(g(x))
 
  override def main(args: Array[String]): Unit = {
    println(compose(makeMultiplier(3), makeAdder(1))(10))
  }
}
